src/main/java/com/epam/Main.java

```diff
@@ -1,24 +1,178 @@
+                        synchronized (filesRead) {filesRead.notify();}
+                    }
+                }));
+
+        synchronized (filesRead) {
+            while(filesRead.intValue() < files.size()) {
+                filesRead.wait();
+            }
+        }
+        readingExecutor.shutdown();
+
+        synchronized (batchesProcessed) {
+            while(batchesProcessed.intValue() < batchesRead.intValue()) {
+                batchesProcessed.wait();
+            }
+        }
+        processingExecutor.shutdown();
+
+        synchronized (batchesWritten) {
+            while(batchesWritten.intValue() < batchesRead.intValue()) {
+                batchesWritten.wait();
+            }
+        }
+        outputExecutor.shutdown();
+    }
+
+    private static int parseIntProp(String propName) {
+        return Integer.valueOf(props.getProperty(propName));
+    }
 
+    private static ExecutorService createExecutor(String name, int maxPoolSize, int workQueueDepth) {
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat(name)
+                .build();
+        return new ThreadPoolExecutor(
+                0, maxPoolSize, 1, TimeUnit.MINUTES,
+                new ArrayBlockingQueue<Runnable>(workQueueDepth),
+                threadFactory, new ThreadPoolExecutor.CallerRunsPolicy());
     }
 
+    private static class ProcessingTask implements Runnable {
+        private DataProcessor processor;
+        private ExecutorService outputExecutor;
+        private List<RoadAccident> accidentsBatch;
+        private AccidentsDataWriter dayWriter;
+        private AccidentsDataWriter nightWriter;
+
+        public ProcessingTask(DataProcessor processor, ExecutorService outputExecutor, List<RoadAccident> accidentsBatch,
+                              AccidentsDataWriter dayWriter, AccidentsDataWriter nightWriter) {
+            this.processor = processor;
+            this.outputExecutor = outputExecutor;
+            this.accidentsBatch = accidentsBatch;
+            this.dayWriter = dayWriter;
+            this.nightWriter = nightWriter;
+        }
+
+        @Override
+        public void run() {
+            processor.enrichWithTimeOfDay(accidentsBatch);
+            processor.enrichWithForceContact(accidentsBatch);
+
+            batchesProcessed.getAndIncrement();
+            synchronized (batchesProcessed) {batchesProcessed.notify();}
+
+            List<RoadAccident> dayAccidents = accidentsBatch.stream()
+                    .filter(roadAccident -> roadAccident.getTimeOfDay() == TimeOfDay.MORNING || roadAccident.getTimeOfDay() == TimeOfDay.AFTERNOON)
+                    .collect(Collectors.toList());
+
+            List<RoadAccident> nightAccidents = accidentsBatch.stream()
+                    .filter(roadAccident -> roadAccident.getTimeOfDay() == TimeOfDay.EVENING || roadAccident.getTimeOfDay() == TimeOfDay.NIGHT)
+                    .collect(Collectors.toList());
+
```

Comment:
Just general recomendation here. I've tought to avoid generic imports with *. When I do it on my project. Better be on save side and name all things in import sepcifically. You can set your IDE to do it for you. ;)